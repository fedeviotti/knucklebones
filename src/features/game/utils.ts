const getColumnByIndex = (index: number) => index % 3;

type CalculatePlayerUpdatedValuesProps = {
  playerValues: number[];
  valueToPlace: number;
  position: number;
};
export const calculatePlayerUpdatedValues = ({
  playerValues,
  valueToPlace,
  position,
}: CalculatePlayerUpdatedValuesProps) => [
  ...playerValues.slice(0, position),
  valueToPlace,
  ...playerValues.slice(position + 1),
];

type CalculateOpponentUpdatedValuesProps = {
  opponentValues: number[];
  valueToPlace: number;
  position: number;
};
export const calculateOpponentUpdatedValues = ({
  opponentValues,
  valueToPlace,
  position,
}: CalculateOpponentUpdatedValuesProps) => [
  ...opponentValues.map((value, i) => {
    if (getColumnByIndex(i) === getColumnByIndex(position)) {
      return (value === valueToPlace ? 0 : value);
    }
    return value;
  }),
];
