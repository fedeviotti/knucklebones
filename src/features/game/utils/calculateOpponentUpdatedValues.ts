const getColumnByIndex = (index: number) => index % 3;

type CalculateOpponentUpdatedValuesProps = {
  values: number[];
  valueToInsert: number;
  position: number;
};
export const calculateOpponentUpdatedValues = ({
  values,
  valueToInsert,
  position,
}: CalculateOpponentUpdatedValuesProps) => [
  ...values.map((value, i) => {
    if (getColumnByIndex(i) === getColumnByIndex(position)) {
      return (value === valueToInsert ? 0 : value);
    }
    return value;
  }),
];
