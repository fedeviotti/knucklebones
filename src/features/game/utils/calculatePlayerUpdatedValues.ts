type CalculatePlayerUpdatedValuesProps = {
  playerValues: number[];
  valueToPlace: number;
  position: number;
};
export const calculatePlayerUpdatedValues = ({
  playerValues,
  valueToPlace,
  position,
}: CalculatePlayerUpdatedValuesProps) => [
  ...playerValues.slice(0, position),
  valueToPlace,
  ...playerValues.slice(position + 1),
];
