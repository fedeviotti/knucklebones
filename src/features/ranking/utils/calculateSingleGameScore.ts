type Props = {
  delta: number | null;
  winner: string | null;
  playerName: string;
};

export const calculateSingleGameScore = ({ delta = 0, winner = "", playerName }: Props) => {
  if (winner === playerName) {
    return ((delta || 0) * 10) + 10;
  }
  return 10;
};
