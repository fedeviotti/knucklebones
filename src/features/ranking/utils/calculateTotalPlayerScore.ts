type Props = {
  delta: number;
};

export const calculateTotalPlayerScore = ({
  delta = 0,
}: Props) => (delta * 10);
