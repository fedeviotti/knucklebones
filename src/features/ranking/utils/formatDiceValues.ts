export const formatDiceValues = (values: string) => {
  const valuesArray = values.split(",");
  return valuesArray.map((e, i) => {
    if ((i + 1) % 3) {
      return `${e},`;
    }
    if (i + 1 === valuesArray.length) {
      return e;
    }
    return `${e}|`;
  }).join("");
};
