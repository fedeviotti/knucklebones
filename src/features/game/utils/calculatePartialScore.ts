import { PARTIAL_SCORE_MAP } from "~/features/game/constant";

type Props = {
  values: number[];
};
export const calculatePartialScore = ({ values }: Props) => PARTIAL_SCORE_MAP.reduce((
  acc,
  columnCells,
  columnIndex,
) => ({
  ...acc,
  [`column${columnIndex + 1}Score`]: columnCells.reduce((
    accumulator,
    cellIndex,
  ) => {
    const occurrences = values
      .filter((v, i) => v !== 0 && columnCells.includes(i))
      .filter((v) => v === values[cellIndex])
      .length || 1;
    return accumulator + ((values[cellIndex] || 0) * occurrences);
  }, 0),
}), ({
  column1Score: 0,
  column2Score: 0,
  column3Score: 0,
}));
