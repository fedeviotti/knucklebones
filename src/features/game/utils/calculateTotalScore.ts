import { calculatePartialScore } from "~/features/game/utils/calculatePartialScore";

type Props = {
  values: number[];
};

export const calculateTotalScore = ({ values }: Props) => {
  const partialScores = calculatePartialScore({ values });

  return partialScores.column1Score + partialScores.column2Score + partialScores.column3Score;
};
