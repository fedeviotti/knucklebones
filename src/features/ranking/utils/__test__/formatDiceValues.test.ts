import { formatDiceValues } from "~/features/ranking/utils/formatDiceValues";

describe("formatDiceValues", () => {
  it("should return the values formatted properly", () => {
    const result = formatDiceValues("0,0,0,0,0,0,0,0,0");
    expect(result).toBe("0,0,0|0,0,0|0,0,0");
  });
});
