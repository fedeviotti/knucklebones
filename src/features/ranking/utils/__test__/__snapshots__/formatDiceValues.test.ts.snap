// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatDiceValues should return the values formatted properly 1`] = `"0,0,0|0,0,0|0,0,0"`;
